samples/src/main/java/okio/samples/Throttler.java

```diff
@@ -0,0 +1,138 @@
+ * desired throughput of any of the Throttlers.
+ *
+ * <p>To clear any throttling, use {@link #clear()} or specify a <code>byteCount</code> of <code>
+ * 0</code> when setting the throughput.
+ */
+public final class Throttler {
+  private static final int NANOS_PER_MILLI = 1000000;
+
+  private double limit = 0.0d;
+  private long last = -1;
+
+  /** Sets the desired max throughput. <code>byteCount</code> of zero to clear. */
+  public void bytesPerTimePeriod(long byteCount, long duration, TimeUnit unit) {
+    if (unit == null || byteCount < 0 || duration <= 0) {
+      throw new IllegalArgumentException(
+          "byteCount=" + byteCount + " duration=" + duration + " unit=" + unit);
+    }
+
+    synchronized (this) {
+      this.limit = (double) byteCount / unit.toNanos(duration);
+    }
+  }
+
+  /** Clears the max throughput. */
+  public void clear() {
+    synchronized (this) {
+      this.limit = 0.0d;
+    }
+  }
+
+  /**
+   * Purposely delays until the caller of the method has taken a sufficient enough time so as not to
+   * exceed the desired throughput.
+   */
+  private void delay(long byteCount) throws InterruptedIOException {
+    // Key: | = request start/end
+    //      * = state change
+    //      r = read
+    //      w = write
+    //      d = delay
+    //      s = wait for synchronization
+    //
+    // Example #2 - 1 thread reading
+    //   time : -1    t1       t2       t3       t4
+    // thread : |- r -|- r -*d-|- r -*d-|- r -*d-|
+    //
+    // Example #2 - 3 threads of reading
+    //   time : -1    t1       t2       t3       t4       t5       t6
+    // thread : |- r -|- r -*- s -------*- d ----|
+    // thread : |- r -*- d ----|- r -*- s -------*- d ----|
+    // thread : |- r -*- s ----*- d ----|- r -*- s -------*- d ----|
+    //
+    // Example #3 - 3 threads of writing
+    //   time : t1       t2       t3       t4       t5       t6
+    // thread : |- w -|- s -------*- d ----*- w -|
+    // thread : |- d ----*- w -|- s -------*- d ----*- w -|
+    // thread : |- s ----*- d ----*- w -|- s -------*- d ----*- w -|
+    //
+    // Example #4 - 2 threads, 1 read, 1 write
+    //   time : t1       t2       t3       t4       t5       t6
+    // thread : |- r -*d-|- r -*s *- d ----|- r -*s-*- d ----|
+    // thread : |- w -|s-*- d ----*- w -|s-*- d ----*- w -|
+
+    synchronized (this) {
+      if (limit == 0.0d) return;
+
+      long required = (long) (byteCount / limit); // Total time required to take
+      long already = System.nanoTime() - last; // Time since last request
+      long delay = required - already; // Time needs to be delayed
+
+      try {
+        if (delay > 0) {
+          Thread.sleep(delay / NANOS_PER_MILLI, (int) (delay % NANOS_PER_MILLI));
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new InterruptedIOException("interrupted");
+      }
+
+      last = System.nanoTime();
```

Comment:
I don’t think we should build up credits if bandwidth is unused.

If you say 60 bytes per second and you have a 5 minute period where you don’t do anything, I don’t think you get a free 300 bytes to send.